Remove torch.functional entries from the doc ignore list

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158581

Options to address the "undocumented python objects":
1. Reference the functions in the .rst via the `torch.functional` namespace. Note that this changes the generated doc filenames / locations for most of these functions!
2. Document these functions by referencing them from the `torch.` namespace instead, in line with common usage. This would also require setting the `__module__` for these functions and moving entries from `torch.functional`'s `__all__` -> `torch`'s `__all__`, which is BC-breaking.
3. Update the .rst files to also document the `torch.functional` forms of these functions, duplicating docs.

This PR takes option (3) above and:
* Removes all 20 `torch.functional` entries from the doc ignore list
* Removes `torch.functional.align_tensors()` entirely, since we don't want to document it.
    * This is technically BC-breaking, although the previous impl simply errored out. This change could be moved to a separate isolated PR for safety.
* Introduces `torch.aliases.md` as a hidden page for the `torch.functional` aliases to the `torch` analogue functions

cc @ezyang @gchanan

jbschlosser committed:
Remove some torch.functional entries from the doc ignore list

[ghstack-poisoned]

jbschlosser committed:
Update on "Remove some torch.functional entries from the doc ignore list"


These are generally referenced elsewhere in the ignore list so removing them doesn't break anything due to how the list works

[ghstack-poisoned]

janeyx99 reviewed (commented):
After this, can you still call them from the torch namespace? like torch.einsum?

jbschlosser committed:
Update on "Remove torch.functional entries from the doc ignore list"


This PR:
* Removes all 20 `torch.functional` entries from the doc ignore list
* References the functions via the `torch.functional` namespace. Note that this changes the generated doc filenames / locations for most of these functions!
* Removes `torch.functional.align_tensors()` entirely, since we don't want to document it.
    * This is technically BC-breaking, although the previous impl simply errored out. This change could be moved to a separate isolated PR for safety.

Alternatives:
1. Document these functions by referencing them from the `torch.` namespace instead, in line with common usage. This would also require setting the `__module__` for these functions and moving entries from `torch.functional`'s `__all__` -> `torch`'s `__all__`, which is BC-breaking.
2. Update the .rst files to also document the `torch.functional` forms of these functions, duplicating docs

Neither of these seemed preferred but insight welcome!

[ghstack-poisoned]

jbschlosser commented:
> After this, can you still call them from the torch namespace? like torch.einsum?

yep sure enough!

jbschlosser commented:
Updated PR to take the approach of documenting both the `torch.functional.*` and `torch.*` forms to avoid issues

janeyx99 reviewed (approved):
I think this is mostly fine! But I'd probably move all the functional ones to their own section so it's not duplicated unexpectedly, not a big deal though

janeyx99 commented:
I do think moving to mikayla's technique with an alias page should be standardized for the long term: https://github.com/pytorch/pytorch/pull/158491

jbschlosser committed:
Update on "Remove torch.functional entries from the doc ignore list"


Options to address the "undocumented python objects":
1. Reference the functions in the .rst via the `torch.functional` namespace. Note that this changes the generated doc filenames / locations for most of these functions!
2. Document these functions by referencing them from the `torch.` namespace instead, in line with common usage. This would also require setting the `__module__` for these functions and moving entries from `torch.functional`'s `__all__` -> `torch`'s `__all__`, which is BC-breaking.
3. Update the .rst files to also document the `torch.functional` forms of these functions, duplicating docs.

This PR takes option (3) above and:
* Removes all 20 `torch.functional` entries from the doc ignore list
* Removes `torch.functional.align_tensors()` entirely, since we don't want to document it.
    * This is technically BC-breaking, although the previous impl simply errored out. This change could be moved to a separate isolated PR for safety.

Neither of these seemed preferred but insight welcome!

[ghstack-poisoned]

jbschlosser committed:
Update on "Remove torch.functional entries from the doc ignore list"


Options to address the "undocumented python objects":
1. Reference the functions in the .rst via the `torch.functional` namespace. Note that this changes the generated doc filenames / locations for most of these functions!
2. Document these functions by referencing them from the `torch.` namespace instead, in line with common usage. This would also require setting the `__module__` for these functions and moving entries from `torch.functional`'s `__all__` -> `torch`'s `__all__`, which is BC-breaking.
3. Update the .rst files to also document the `torch.functional` forms of these functions, duplicating docs.

This PR takes option (3) above and:
* Removes all 20 `torch.functional` entries from the doc ignore list
* Removes `torch.functional.align_tensors()` entirely, since we don't want to document it.
    * This is technically BC-breaking, although the previous impl simply errored out. This change could be moved to a separate isolated PR for safety.
* Introduces `torch.aliases.md` as a hidden page for the `torch.functional` aliases to the `torch` analogue functions

[ghstack-poisoned]

janeyx99 reviewed (approved):
approving again, wow, so systematic now that we figured it out

jbschlosser commented:
@pytorchbot merge -i
